src/analyses/mutexAnalysis.ml

```diff
@@ -23,92 +20,42 @@ end
 (** Mutex analyzer without base --- this is the new standard *)
 module MakeSpec (P: SpecParam) =
 struct
-  include Analyses.DefaultSpec
-
-  (** name for the analysis (btw, it's "Only Mutex Must") *)
+  module Arg =
+  struct
+    module D = Lockset
+    module G = P.G
+    module V = VarinfoV
+
+    let add ctx l =
+      D.add l ctx.local
+
+    let remove ctx l =
+      D.remove (l, true) (D.remove (l, false) ctx.local)
+
+    let remove_all ctx =
+      (* TODO: implement in privatizations? *)
+      (* Mutexes.iter (fun m ->
+          ctx.emit (MustUnlock m)
+        ) (D.export_locks ctx.local); *)
+      (* TODO: used to have remove_nonspecial, which kept v.vname.[0] = '{' variables *)
+      D.empty ()
+  end
+  include LocksetAnalysis.MakeMust (Arg)
   let name () = "mutex"
 
-  (** Add current lockset alongside to the base analysis domain. Global data is collected using dirty side-effecting. *)
-  module D = Lockset
-  module C = Lockset
+  module D = Arg.D (* help type checker using explicit constraint *)
+  let should_join x y = D.equal x y
 
-  (** We do not add global state, so just lift from [BS]*)
+  (** Global data is collected using dirty side-effecting. *)
   module G = P.G
   module V = VarinfoV
 
-  let should_join x y = D.equal x y
-
-  (* NB! Currently we care only about concrete indexes. Base (seeing only a int domain
-     element) answers with the string "unknown" on all non-concrete cases. *)
-  let rec conv_offset x =
-    match x with
-    | `NoOffset    -> `NoOffset
-    | `Index (Const (CInt (i,_,s)),o) -> `Index (IntDomain.of_const (i,Cilfacade.ptrdiff_ikind (),s), conv_offset o)
-    | `Index (_,o) -> `Index (ValueDomain.IndexDomain.top (), conv_offset o)
-    | `Field (f,o) -> `Field (f, conv_offset o)
-
   let rec conv_offset_inv = function
```

Comment:
This is duplicated now?

Reply:
Indeed, it's unused in MutexEventsAnalysis, so I'll remove it from there.